File: HealthyDay-OC/HealthManager.m
//
//  HealthManager.m
//  HealthyDay
//
//  Created by Linsw on 16/11/19.
//  Copyright © 2016年 Linsw. All rights reserved.
//

#import "HealthManager.h"
#import "Define.h"

@interface HealthManager()
@property HKHealthStore *store;

@end


@implementation HealthManager

+(instancetype)sharedHealthManager {
    static HealthManager *sharedHealthManager = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken,^{
        sharedHealthManager = [[self alloc] init];
    });
    return sharedHealthManager;
}

- (id)init{
    self = [super init];
    if (self){
        self.store = [[HKHealthStore alloc] init];
    }
    return self;
}

#pragma mark Public API
-(void)authorizeWithCompletion:(void(^)(BOOL,NSError* _Nullable))completion{
    NSSet *typesToRead = [NSSet setWithArray:@[
                                               [HKQuantityType quantityTypeForIdentifier:HKQuantityTypeIdentifierStepCount],
                                               [HKQuantityType quantityTypeForIdentifier:HKQuantityTypeIdentifierDistanceWalkingRunning]
                                               ]];
    if (![HKHealthStore isHealthDataAvailable]){
//        NSError *error = [NSError errorWithDomain:[@"com.linshiwei.healthkit" code:2 userInfo:@{NSLocalizedDescriptionKey:@"HealthKit is not available in this Device"}];
        completion(NO,nil);
        return;
    }
    [_store requestAuthorizationToShareTypes:nil readTypes:typesToRead completion:^(BOOL success, NSError *error){
        completion(success,error);
    }];
    
    
}

-(void)readStepCountInDate:(NSDate*)date withPeriodDataType:(enum PeriodDataType)type withCompletion:(void(^)(NSArray<NSNumber *> * _Nullable,NSError* _Nullable))completion{
    __block NSArray<NSNumber *> *stepCounts;
    HKSampleQuery *sampleQuery = [self createQuantitySampleQueryInDate:date typeIdentifier:HKQuantityTypeIdentifierStepCount periodDataType:type withCompletion:^(HKSampleQuery *query,NSArray<HKSample *> *samples,NSError *error){
        if (error == nil && samples != nil) {
            stepCounts = [self calculatePerDayDataFromAscendingSamples:samples typeIdentifier:HKQuantityTypeIdentifierStepCount periodDataType:type];
            completion(stepCounts,nil);
        }else{
            completion(nil,error);
        }
    }];
    [_store executeQuery:sampleQuery];
}

-(void)readStepCountWithPeriodDataType:(enum PeriodDataType)type withCompletion:(void(^)(NSArray<NSNumber *> * _Nullable,NSError* _Nullable))completion{
    [self readStepCountInDate:[NSDate date] withPeriodDataType:type withCompletion:completion];
}

-(void)readDistanceWalkingRunningInDate:(NSDate*)date withPeriodType:(enum PeriodDataType)type withCompletion:(void(^)(NSArray<NSNumber *> * _Nullable,NSError* _Nullable))completion{
    __block NSArray<NSNumber *> *distances;
    HKSampleQuery * sampleQuery = [self createQuantitySampleQueryInDate:date typeIdentifier:HKQuantityTypeIdentifierDistanceWalkingRunning periodDataType:type withCompletion:^(HKSampleQuery *query,NSArray<HKSample *> * samples,NSError * error){
        if (error == nil && samples != nil) {
            distances = [self calculatePerDayDataFromAscendingSamples:samples typeIdentifier:HKQuantityTypeIdentifierDistanceWalkingRunning periodDataType:type];
            completion(distances,nil);
        }else{
            completion(nil,error);

        }
    }];
    [_store executeQuery:sampleQuery];
}

-(void)readDistanceWalkingRunningWithPeriodType:(enum PeriodDataType)type withCompletion:(void(^)(NSArray<NSNumber *> * _Nullable,NSError* _Nullable))completion{
    [self readDistanceWalkingRunningInDate:[NSDate date] withPeriodType:type withCompletion:completion];
}


-(void)readDetailStepCountInDate:(NSDate*)date withCompletion:(void(^)(NSArray<HKSample *> * _Nullable,NSError* _Nullable))completion{
    HKSampleQuery *detailSampleQuery = [self createQuantitySampleQueryInDate:date typeIdentifier:HKQuantityTypeIdentifierStepCount periodDataType:Specified withCompletion:^(HKSampleQuery *query,NSArray<HKSample *> *samples,NSError *error){
        if (error == nil && samples != nil) {
            completion(samples,nil);
        }else{
            completion(nil,error);
        }
    }];
    [_store executeQuery:detailSampleQuery];
}


#pragma mark Private Help Func
- (HKSampleQuery *)createQuantitySampleQueryInDate:(NSDate*)date typeIdentifier:(HKQuantityTypeIdentifier)identifier periodDataType:(enum PeriodDataType)type withCompletion:(void(^)(HKSampleQuery*,NSArray<HKSample *> * _Nullable,NSError* _Nullable))completion{
    NSDate *beginningDate = [self getBeginningDateWithPeriodDataType:type ofDate:date];
    NSDate *endDate = (type == Specified)? [beginningDate dateByAddingTimeInterval:24*3600] : date;
    HKQuantityType *readingType = [HKQuantityType quantityTypeForIdentifier:identifier];
    NSPredicate *currentDayPerdicate = [HKQuery predicateForSamplesWithStartDate:beginningDate endDate:endDate options:HKQueryOptionNone];
    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:HKSampleSortIdentifierStartDate ascending:true];
    return [[HKSampleQuery alloc] initWithSampleType:readingType predicate:currentDayPerdicate limit:HKObjectQueryNoLimit sortDescriptors:[NSArray arrayWithObject:sortDescriptor] resultsHandler:completion];
}

- (NSArray<NSNumber *>*)calculatePerDayDataFromAscendingSamples:(NSArray<HKSample *> *)samples typeIdentifier:(HKQuantityTypeIdentifier)identifier periodDataType:(enum PeriodDataType)type{
    HKUnit *unit = [HKUnit countUnit];
    if (identifier == HKQuantityTypeIdentifierStepCount){
        unit = [HKUnit countUnit];
    }else{
        if (identifier == HKQuantityTypeIdentifierDistanceWalkingRunning) {
            unit = [HKUnit meterUnit];
        }else{
            
        }
    }
    NSMutableArray *counts;
    int count = 0;
    NSDate *perDay = [NSDate dateWithTimeInterval:24*3600 sinceDate:[[Define sharedDefine].calendar startOfDayForDate:[samples[0] endDate]]];
    for (HKSample *sample in samples){
        if ([[sample endDate] compare:perDay] == NSOrderedAscending){
            count += [[(HKQuantitySample*)sample quantity] doubleValueForUnit:unit];
        }else{
            NSTimeInterval interval = [[sample endDate] timeIntervalSinceDate:[sample startDate]];
            if ([[[sample startDate] dateByAddingTimeInterval:interval] compare:perDay] == NSOrderedAscending) {
                count += [[(HKQuantitySample*)sample quantity] doubleValueForUnit:unit];
                [counts addObject:[NSNumber numberWithInt:count]];
                count = 0;
            }else{
                [counts addObject:[NSNumber numberWithInt:count]];
                count = [[(HKQuantitySample*)sample quantity] doubleValueForUnit:unit];
            }
            perDay = [NSDate dateWithTimeInterval:24*3600 sinceDate:perDay];
        }
    }
    [counts addObject:[NSNumber numberWithInt:count]];
    switch (type) {
        case Current:
        case Specified:
            NSAssert([counts count] <= 1,@"counts.count <=1");
            break;
        case Weekly:
            NSAssert([counts count] <= 7,@"counts.count <=7");
            break;
        case Monthly:
            NSAssert([counts count] <= 30,@"counts.count <=30");
            break;
        default:
            break;
    }
    return counts;
}

- (NSDate*)getBeginningDateWithPeriodDataType:(enum PeriodDataType)type ofDate:(NSDate*)date{
    NSCalendar *calendar = [Define sharedDefine].calendar;
    switch (type) {
        case Current:
            return [calendar startOfDayForDate:date];
            break;
        case Specified:
            return [calendar startOfDayForDate:date];
            break;
        case Weekly:
            return [calendar dateByAddingUnit:NSCalendarUnitDay value:-6 toDate:date options:NSCalendarMatchStrictly];
        case Monthly:
            return [calendar dateByAddingUnit:NSCalendarUnitDay value:-6 toDate:date options:NSCalendarMatchStrictly];
        default:
            break;
    }
}
@end

File: HealthyDay/HealthManager.swift
//
//  HealthManager.swift
//  HealthyDay
//
//  Created by Linsw on 16/9/26.
//  Copyright © 2016年 Linsw. All rights reserved.
//

import Foundation
import HealthKit

enum PeriodDataType{
    case Specified
    case Current
    case Weekly
    case Monthly
}

internal final class HealthManager{
    static let sharedHealthManager = HealthManager()
    private let store = HKHealthStore()
    
    private init(){
    }
//MARK: API
    internal func authorize(_ completion: @escaping (_ success:Bool, _ error:Error?) -> Void){
        let typesToRead : Set = [
            HKQuantityType.quantityType(forIdentifier: .stepCount)!,
            HKQuantityType.quantityType(forIdentifier: .distanceWalkingRunning)!
            ]
        
        if !HKHealthStore.isHealthDataAvailable(){
            completion(false, nil)
            return
        }
        store.requestAuthorization(toShare: nil, read: typesToRead){(success, error) in
            completion(success,error)
        }
    }
    
    internal func readStepCount(inDate date:Date = Date(), periodDataType type:PeriodDataType, _ completion: @escaping ([Int]?,Error?)->Void){
        var stepCounts = [Int]()
        let sampleQuery = createQuantitySampleQuery(inDate: date, typeIdentifier: .stepCount, periodDataType: type){[unowned self](query, samples, error) in
            guard error == nil && samples != nil else{
                completion(nil,error)
                return
            }
            stepCounts = self.calculatePerDayData(fromAscendingSamples: samples!, typeIdentifier:.stepCount, periodDataType: type)
            completion(stepCounts,nil)
        }
        store.execute(sampleQuery)
    }
    
    internal func readDistanceWalkingRunning(inDate date:Date = Date(), periodDataType type:PeriodDataType, _ completion: @escaping ([Int]?,Error?)->Void){
        var distances = [Int]()
        let sampleQuery = createQuantitySampleQuery(inDate: date, typeIdentifier: .distanceWalkingRunning, periodDataType: type){[unowned self](query, samples, error) in
            guard error == nil && samples != nil else{
                completion(nil,error)
                return
            }
            distances = self.calculatePerDayData(fromAscendingSamples: samples!, typeIdentifier:.distanceWalkingRunning, periodDataType: type)
            completion(distances,nil)
        }
        store.execute(sampleQuery)
    }
    
    internal func readDetailStepCount(inDate date:Date, _ completion: @escaping ([HKSample]?,Error?)->Void){

        let detailSampleQuery = createQuantitySampleQuery(inDate: date, typeIdentifier: .stepCount, periodDataType: .Specified){(query, samples, error) in
            guard error == nil && samples != nil else{
                completion(nil,error)
                return
            }
            completion(samples,nil)
        }
        store.execute(detailSampleQuery)
    }
//MARK: Help func
    private func createQuantitySampleQuery(inDate date:Date = Date(), typeIdentifier identifier:HKQuantityTypeIdentifier, periodDataType type:PeriodDataType, completion:  @escaping (HKSampleQuery,[HKSample]?,Error?)->Void)->HKSampleQuery{
        
        let beginningDate = getBeginningDate(withPeriodDataType: type, ofDate: date)
        let endDate = type == .Specified ? beginningDate.addingTimeInterval(3600*24) : date
        let readingType = HKQuantityType.quantityType(forIdentifier: identifier)!
        let currentDayPredicate = HKQuery.predicateForSamples(withStart: beginningDate, end: endDate, options: HKQueryOptions())
        let sortDescriptor = NSSortDescriptor(key: HKSampleSortIdentifierStartDate, ascending: true)
        return HKSampleQuery(sampleType: readingType, predicate: currentDayPredicate, limit: HKObjectQueryNoLimit, sortDescriptors: [sortDescriptor], resultsHandler: completion)
    
    }
    
    private func calculatePerDayData(fromAscendingSamples samples:[HKSample], typeIdentifier identifier:HKQuantityTypeIdentifier, periodDataType type:PeriodDataType)->[Int]{
        let unit : HKUnit = {
            switch identifier {
            case HKQuantityTypeIdentifier.stepCount:
                return HKUnit.count()
            case HKQuantityTypeIdentifier.distanceWalkingRunning:
                return HKUnit.meter()
            default:
                fatalError("Unsupport quantity type")
            }
        }()
        var counts = [Int]()
        var count = 0
        var perDay = Date(timeInterval: 24*3600, since: calendar.startOfDay(for: samples[0].endDate))
        
        for sample in samples{
            if sample.endDate <= perDay {
                count += Int((sample as! HKQuantitySample).quantity.doubleValue(for: unit))
            }else{
                let interval = sample.endDate.timeIntervalSince(sample.startDate)
                if sample.startDate+interval/2 <= perDay {
                    count += Int((sample as! HKQuantitySample).quantity.doubleValue(for: unit))
                    counts.append(count)
                    count = 0
                }else{
                    counts.append(count)
                    count = Int((sample as! HKQuantitySample).quantity.doubleValue(for: unit))
                }
                perDay = Date(timeInterval: 24*3600, since: perDay)
            }
        }
        
        counts.append(count)
        switch type {
        case .Current,.Specified:
            assert(counts.count <= 1)
        case .Weekly:
            assert(counts.count <= 7)
        case .Monthly:
            assert(counts.count <= 30)
        }
        return counts
    }

    private func getBeginningDate(withPeriodDataType type:PeriodDataType, ofDate date:Date)->Date{
        switch type {
        case .Current,.Specified:
            return calendar.startOfDay(for: date)
        case .Weekly:
            return calendar.date(byAdding: .day, value: -6, to: calendar.startOfDay(for: date))!
        case .Monthly:
            return calendar.date(byAdding: .day, value: -29, to: calendar.startOfDay(for: date))!
        }
    }
}
